### ISSUE TITLE
"couldn't resolve module/action" for local module used in role

### ISSUE THREAD

**Issue by jwodder** [ansible/ansible-lint#3846]
I have an Ansible playbook laid out as follows:

```text
./
├── hosts
├── library/
│   └── nop.py
├── playbook.yml
└── roles/
    └── myrole/
        └── tasks/
            └── main.yml
```

Contents of `library/nop.py`:

```python
#!/usr/bin/python3
from ansible.module_utils.basic import AnsibleModule

DOCUMENTATION = r"""
---
module: nop
short_description: Does nothing
options:
"""


def run_module():
    module = AnsibleModule(argument_spec={}, supports_check_mode=True)
    module.exit_json(changed=False)


def main():
    run_module()


if __name__ == "__main__":
    main()
```

Contents of `playbook.yml`:

```yaml
---
- name: Play
  hosts: all
  roles:
    - myrole
```

Contents of `roles/myrole/tasks/main.yml`:

```yaml
---
- name: Do nothing
  nop:
```

Running `ansible-lint` on this directory gives the error:

```text
WARNING  Listing 1 violation(s) that are fatal
syntax-check[specific]: couldn't resolve module/action 'nop'. This often indicates a misspelling, missing collection, or incorrect module path.
roles/myrole/tasks/main.yml:2:3


                  Rule Violation Summary                   
 count tag                    profile rule associated tags 
     1 syntax-check[specific] min     core, unskippable    

Failed: 1 failure(s), 0 warning(s) on 5 files.
```

However, if the `nop` module is invoked as a task directly in `playbook.yml` instead, `ansible-lint` is fine with it.

##### OS / ENVIRONMENT

```console
ansible-lint 6.20.3 using ansible-core:2.15.5 ansible-compat:4.1.10 ruamel-yaml:0.17.35 ruamel-yaml-clib:0.2.8
```

- ansible installation method: pip
- ansible-lint installation method: pip

**Comment by mshonichev** [comment]
anyone got simple workaround?

**Comment by depate** [comment]
I am running into the same error, however, only whilst using `ansible-lint@main` as a GitHub workflow action. Running `ansible-lint` locally did not fail, yet.

The used version locally is:
```❯ ansible-lint --version
ansible-lint 6.22.1 using ansible-core:2.16.2 ansible-compat:4.1.10 ruamel-yaml:0.17.40 ruamel-yaml-clib:0.2.7```

**Comment by vbotka** [comment]
I see this error when importing a role into the Ansible Galaxy. The importer ignores this problem.

https://galaxy.ansible.com/ui/standalone/roles/vbotka/ansible_lint/import_log/

```
===== LOADING ROLE ===== 
Importing with galaxy-importer 0.4.19 
Determined role name to be ansible_lint 
Linting role ansible_lint via ansible-lint... 
ansible-lint/tasks/packages.yml:9:7: syntax-check[specific]: couldn't resolve module/action 'community.general.pkgng'. This often indicates a misspelling, missing collection, or incorrect module path. 
...ansible-lint run complete 
Legacy role loading complete 
```

I see this problem on all my roles that use any collection. For example,

https://galaxy.ansible.com/ui/standalone/roles/vbotka/linux_postinstall/import_log/

Other "heavy-duty" roles suffer from this too. For example,

https://galaxy.ansible.com/ui/standalone/roles/mrlesmithjr/netplan/import_log/

**Comment by plusiv** [comment]
> I am running into the same error, however, only whilst using `ansible-lint@main` as a GitHub workflow action. Running `ansible-lint` locally did not fail, yet.
> 
> The used version locally is:
> 
> ```
> ❯ ansible-lint --version
> ansible-lint 6.22.1 using ansible-core:2.16.2 ansible-compat:4.1.10 ruamel-yaml:0.17.40 ruamel-yaml-clib:0.2.7
> ```
> 
> Edit:
> 
> Maybe my issue is not related. Adding the module/collection in `requirements.yml` solved the issue.

Worked for me adding `requirements.yml` in the root folder.

> [!NOTE]  
> For some reason it does not work with `requirements.yaml`. Seems like is not supported in the [documentation](https://docs.ansible.com/ansible/latest/galaxy/user_guide.html#installing-roles-and-collections-from-the-same-requirements-yml-file).

**Comment by vbotka** [comment]
*requirements.yml* in the root folder does not help

```yaml
shell> cat requirements.yml 
---
collections:
  - name: community.general
    source: https://galaxy.ansible.com
```

Galaxy reports the same error:

```
===== LOADING ROLE ===== 
Importing with galaxy-importer 0.4.20 
Determined role name to be ansible_lint 
Linting role ansible_lint via ansible-lint... 
ansible-lint/tasks/packages.yml:9:7: syntax-check[specific]: couldn't resolve module/action 'community.general.pkgng'. This often indicates a misspelling, missing collection, or incorrect module path. 
...ansible-lint run complete 
Legacy role loading complete
```

See:
https://galaxy.ansible.com/ui/standalone/roles/vbotka/ansible_lint/import_log/

**Comment by anl** [comment]
A workaround which addresses the related issue described in #4042 (closed as duplicate of this issue) is to set the `ANSIBLE_LIBRARY` environment variable appropriately.

For example, given a role laid out like this:

```
.
├── defaults
│   └── main.yml
├── library
│   └── mongodb_replicaset_members.py
├── meta
│   └── main.yml
├── tasks
│   └── main.yml
└── vars
    └── main.yml
```

Running `ansible-lint` in the top directory fails with:

```
syntax-check[unknown-module]: couldn't resolve module/action 'mongodb_replicaset_members'. This often indicates a misspelling, missing collection, or incorrect module path.
```

However, running `ANSIBLE_LIBRARY=library ansible-lint` succeeds.

Tested with:
```
$ ansible-lint --version
ansible-lint 24.2.1 using ansible-core:2.16.5 ansible-compat:4.1.11 ruamel-yaml:0.18.6 ruamel-yaml-clib:0.2.8
```

**Comment by jwodder** [comment]
This is still happening with ansible-lint 25.11.0, only now the error type is `syntax-check[unknown-module]`.

**Comment by vbotka** [comment]
> This is still happening ...

See also https://galaxy.ansible.com/ui/standalone/roles/vbotka/linux_postinstall/import_log/

```
===== LOADING ROLE ===== 
Importing with galaxy-importer 0.4.31 
Determined role name to be linux_postinstall 
Linting role linux_postinstall via ansible-lint 24.12.2... 
  ...
ansible-linux-postinstall/tasks/swap.yml:84:7: syntax-check[unknown-module]: couldn't resolve module/action 'ansible.posix.mount'. This often indicates a misspelling, missing collection, or incorrect module path. 
```

For years, it has been evident in all roles. For example:

* https://galaxy.ansible.com/ui/standalone/roles/vbotka/freebsd_zfs/import_log/

```
ansible-freebsd-zfs/tasks/rcconf.yml:3:3: syntax-check[unknown-module]: couldn't resolve module/action 'community.general.sysrc'. This often indicates a misspelling, missing collection, or incorrect module path. 
```

* https://galaxy.ansible.com/ui/standalone/roles/vbotka/freebsd_wpa_cli/import_log/

```
ansible-freebsd-wpa-cli/tasks/packages.yml:2:3: syntax-check[unknown-module]: couldn't resolve module/action 'community.general.pkgng'. This often indicates a misspelling, missing collection, or incorrect module path. 
```
